system_metrics_collector/test/system_metrics_collector/test_linux_cpu_measurement.cpp

```diff
@@ -54,16 +102,100 @@ class TestLinuxCpuMeasurementNode : public system_metrics_collector::LinuxCpuMea
+  : rclcpp::Node(name), times_received(0)
+  {
+    auto callback = [this](MetricsMessage::UniquePtr msg) {this->MetricsMessageCallback(*msg);};
+    subscription = create_subscription<MetricsMessage,
+        std::function<void(MetricsMessage::UniquePtr)>>(TEST_TOPIC, 10, callback);
+
+    for (auto & stats : expected_stats) {
+      for (int i = 0; i < STATISTICS_DATA_TYPES.size(); ++i) {
+        stats[i].data_type = STATISTICS_DATA_TYPES[i];
+        stats[i].data = std::nan("");
+      }
     }
+
+    // setting expected_stats[0]
+    expected_stats[0][StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT - 1].data = 0;
+
+    // setting expected_stats[1]
+    moving_average_statistics::MovingAverageStatistics stats_calc;
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[0], proc_samples[1]));
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[1], proc_samples[2]));
+    StatisticData data = stats_calc.getStatistics();
+    StatisticDataToStatisticDataPoints(data, expected_stats[1]);
+
+    // setting expected_stats[2]
+    stats_calc.reset();
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[2], proc_samples[3]));
+    data = stats_calc.getStatistics();
+    StatisticDataToStatisticDataPoints(data, expected_stats[2]);
+
+    // setting expected_stats[3]
+    expected_stats[3][StatisticDataType::STATISTICS_DATA_TYPE_SAMPLE_COUNT - 1].data = 0;
+
+    // setting expected_stats[4]
+    stats_calc.reset();
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[5], proc_samples[6]));
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[6], proc_samples[7]));
+    data = stats_calc.getStatistics();
+    StatisticDataToStatisticDataPoints(data, expected_stats[4]);
+
+    // setting expected_stats[5]
+    stats_calc.reset();
+    stats_calc.addMeasurement(computeCpuActivePercentage(proc_samples[7], proc_samples[8]));
+    data = stats_calc.getStatistics();
+    StatisticDataToStatisticDataPoints(data, expected_stats[5]);
+  }
+
+  int getNumReceived() const
+  {
+    return times_received;
   }
 
-  bool first{false};
+private:
+  void MetricsMessageCallback(const MetricsMessage & msg)
+  {
+    ASSERT_GT(expected_stats.size(), times_received);
+
+    // check source names
+    EXPECT_EQ(TEST_NODE_NAME, msg.measurement_source_name);
+    EXPECT_EQ("cpu_usage", msg.metrics_source);
+
+    // check measurement window
+    // std::chrono::seconds window_sec(msg.window_stop.sec - msg.window_start.sec);
+    // std::chrono::nanoseconds window_nanosec(msg.window_stop.nanosec - msg.window_start.nanosec);
+    // std::chrono::milliseconds window = std::chrono::duration_cast<std::chrono::milliseconds>(
+    //   window_sec + window_nanosec);
+    // EXPECT_GT(5, std::abs(window.count() - PUBLISH_PERIOD.count()));
+
+    // check measurements
+    for (int i = 0; i < STATISTICS_DATA_TYPES.size(); ++i) {
```

Comment:
Could we avoid having i used in two arrays of potentially different sizes?

Reply:
Done. One of the two original arrays is now a `std::unordered_map`.